radicle-node/src/wire/protocol.rs

```diff
@@ -0,0 +1,315 @@
+                self.disconnect(
+                    fd,
+                    DisconnectReason::Protocol(service::DisconnectReason::Error(
+                        session::Error::Misbehavior,
+                    )),
+                );
+            }
+            SessionEvent::ConnectionFailure(err) => {
+                self.disconnect(fd, DisconnectReason::ConnectionError(Arc::new(err)));
+            }
+            SessionEvent::Disconnected => {
+                self.disconnect(
+                    fd,
+                    DisconnectReason::Protocol(service::DisconnectReason::Peer),
+                );
+            }
+        }
+    }
+
+    fn handle_command(&mut self, cmd: Self::Command) {
+        self.inner.command(cmd)
+    }
+
+    fn handle_error(&mut self, err: reactor::Error<net::SocketAddr, RawFd>) {
+        match err {
+            reactor::Error::ListenerUnknown(id) => {
+                log::error!("asking for unknown listener {}", id);
+            }
+            reactor::Error::PeerUnknown(id) => {
+                log::error!("asking for unknown peer {}", id);
+            }
+            reactor::Error::PeerDisconnected(addr, err) => {
+                log::error!("the peer {} got disconnected with {}", addr, err);
+                self.inner_queue
+                    .push_back(reactor::Action::UnregisterTransport(addr))
+            }
+        }
+    }
+
+    fn handover_listener(&mut self, _listener: Self::Listener) {
+        // TODO(cloudhead): What should we do here?
+        unreachable!("for now we do not disconnect socket we listening on")
+    }
+
+    fn handover_transport(&mut self, transport: Self::Transport) {
+        let fd = transport.as_raw_fd();
+        if let Some(reason) = self.hangups.get(&fd) {
+            if let Some(reason) = reason {
+                self.inner.disconnected(transport.expect_peer_id(), reason);
+            }
+        } else {
+            todo!("send transport to the worker")
+        }
+    }
+}
+
+impl<R, S, W, G, N> Iterator for Wire<R, S, W, G, N>
+where
+    R: routing::Store,
+    S: address::Store,
+    W: WriteStorage + 'static,
+    G: Signer,
+    N: Negotiator,
+{
+    type Item = reactor::Action<NetAccept<Session<N>>, NetTransport<Session<N>, Message>>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some(event) = self.inner_queue.pop_front() {
+            return Some(event);
+        }
+
+        while let Some(ev) = self.inner.next() {
+            match ev {
+                Io::Write(node_id, msgs) => {
+                    log::debug!("Sending {} messages to {}", msgs.len(), node_id);
+                    let (fd, _) = self
+                        .sessions
+                        .iter()
+                        .find(|(_, id)| **id == node_id)
+                        .expect("sending message to the peer which is not connected");
+                    return Some(reactor::Action::Send(*fd, msgs));
```

Comment:
Could we rename this `Send` variant? It's confusing for syntax highlighters :) Maybe `SendMessage`?

Reply:
Yup, either works for me.